File: dataset/generate.py
# -*- coding: utf-8 -*-
import json
import random

prefixes = ["sweet ", "red ", "green ", "winter ", "summer ", "wild ", "hybrid ", "northern ", "southern ", "golden ", "black ", "white ", "blue ", "organic ", "mountain ", "valley "]
bases = [
    "wheat", "rice", "corn", "bean", "pea", "potato", "tomato", "cotton", "soybean", "barley", 
    "oat", "sorghum", "millet", "rye", "sugarcane", "yam", "cassava", "pepper", "onion", "garlic", 
    "carrot", "radish", "turnip", "cabbage", "lettuce", "spinach", "broccoli", "cauliflower", 
    "pumpkin", "squash", "cucumber", "melon", "watermelon", "berry", "strawberry", "blueberry", 
    "raspberry", "blackberry", "grape", "apple", "pear", "peach", "plum", "cherry", "apricot", 
    "mango", "banana", "citrus", "orange", "lemon", "lime", "grapefruit", "almond", "walnut", 
    "pecan", "pistachio", "cashew", "peanut", "sunflower", "safflower", "sesame", "flax", "hemp", 
    "clover", "alfalfa", "timothy", "fescue", "vetch", "lupine", "chickpea", "lentil", "mustard", 
    "canola", "olive", "date", "fig", "pomegranate", "kiwi", "papaya", "guava", "pineapple", 
    "coconut", "cocoa", "coffee", "tea", "vanilla", "cinnamon", "nutmeg", "clove", "ginger", 
    "turmeric", "cardamom", "pepper", "mint", "basil", "oregano", "thyme", "rosemary", "sage", 
    "lavender", "hops", "tobacco", "jute", "sisal", "rubber", "bamboo", "rattan", "gourd", "taro", 
    "artichoke", "asparagus", "celery", "eggplant", "okra"
]

seasons = ["Kharif", "Rabi", "Zaid", "Annual", "Perennial"]
icons_pool = [
    "fa-sun", "fa-cloud-rain", "fa-snowflake", "fa-cloud-meatball", 
    "fa-cloud-showers-heavy", "fa-leaf", "fa-mug-hot", "fa-temperature-arrow-up", 
    "fa-water", "fa-apple-whole", "fa-seedling", "fa-tree", "fa-wind", "fa-droplet"
]

original = {
    "rice": { 
        "temp": "22°C - 32°C", "rain": "150mm - 300mm", "humidity": "70% - 80%", "season": "Kharif", "icons": ["fa-sun", "fa-cloud-rain"],
        "scientific_name": "Oryza sativa",
        "varieties": ["Basmati", "Jasmine", "Arborio"],
        "info": "Rice is a staple food for over half the world's population. It thrives in waterlogged conditions and clay-rich soil.",
        "climate_needs": "Requires high humidity, prolonged sunshine, and an assured supply of water.",
        "ideal_conditions": "Heavy clay or clay loam soils that can retain water, mainly in tropical and subtropical regions."
    },
    "wheat": { 
        "temp": "10°C - 25°C", "rain": "50mm - 100mm", "humidity": "50% - 60%", "season": "Rabi", "icons": ["fa-snowflake", "fa-sun"],
        "scientific_name": "Triticum aestivum",
        "varieties": ["Sharbati", "Durum", "Spelt"],
        "info": "Wheat is a globally cultivated grass grown for its seed. It prefers well-drained loamy to clay loamy soils.",
        "climate_needs": "Requires cool, moist conditions during growth and warm, dry weather for ripening.",
        "ideal_conditions": "Well-drained fertile loamy and clayey soils; widely grown in temperate and subtropical areas."
    },
    "cotton": { 
        "temp": "21°C - 30°C", "rain": "50mm - 100mm", "humidity": "40% - 60%", "season": "Kharif", "icons": ["fa-sun"],
        "scientific_name": "Gossypium",
        "varieties": ["Pima", "Upland", "Egyptian"],
        "info": "Cotton is a soft, fluffy staple fiber that grows in a boll. It requires long frost-free periods and plenty of sunshine.",
        "climate_needs": "Needs frost-free days, high sunshine hours, and warm temperatures during active growth.",
        "ideal_conditions": "Deep, well-drained black soil or alluvial soils with good moisture retention."
    },
    "maize": { 
        "temp": "21°C - 27°C", "rain": "50mm - 100mm", "humidity": "55% - 75%", "season": "Kharif", "icons": ["fa-sun", "fa-cloud-meatball"],
        "scientific_name": "Zea mays",
        "varieties": ["Sweet Corn", "Flint Corn", "Dent Corn"],
        "info": "Maize, or corn, is a cereal grain known for its versatile uses. It needs deep, rich, well-draining soils with ample nitrogen.",
        "climate_needs": "Requires warm weather with evenly distributed rainfall throughout its growing season.",
        "ideal_conditions": "Well-drained, deep, fertile loam or silty loam soils with a neutral pH."
    },
    "sugarcane": { 
        "temp": "21°C - 27°C", "rain": "150mm - 250mm", "humidity": "70% - 85%", "season": "Annual", "icons": ["fa-cloud-rain", "fa-sun"],
        "scientific_name": "Saccharum officinarum",
        "varieties": ["CO-0238", "CO-86032"],
        "info": "Sugarcane is a tall, perennial grass used for sugar production. It requires a long, warm growing season with high moisture.",
        "climate_needs": "Demands high temperatures, bright sunlight, and frequent rainfall or irrigation.",
        "ideal_conditions": "Deep, well-drained rich loamy soils, particularly near tropical and subtropical coastlines."
    },
    "tea": { 
        "temp": "20°C - 30°C", "rain": "150mm - 300mm", "humidity": "70% - 90%", "season": "Perennial", "icons": ["fa-cloud-showers-heavy", "fa-leaf"],
        "scientific_name": "Camellia sinensis",
        "varieties": ["Assam", "Darjeeling", "Green"],
        "info": "Tea plants are evergreen shrubs prized for their leaves. They thrive in acidic soils and well-distributed rainfall.",
        "climate_needs": "Needs a warm and humid environment, without stagnant water at the roots or frost.",
        "ideal_conditions": "Deep, well-drained, porous acidic soils, commonly situated on hill slopes."
    },
    "coffee": { 
        "temp": "15°C - 28°C", "rain": "150mm - 250mm", "humidity": "70% - 80%", "season": "Perennial", "icons": ["fa-cloud-rain", "fa-mug-hot"],
        "scientific_name": "Coffea",
        "varieties": ["Arabica", "Robusta"],
        "info": "Coffee plants blossom with white flowers before yielding coffee cherries. They favor slightly acidic, well-drained volcanic soil.",
        "climate_needs": "Requires moderate temperature and rainfall, shaded by larger canopy trees often.",
        "ideal_conditions": "Rich, well-drained loamy soils enriched with humus, primarily in highland tropical regions."
    },
    "jute": { 
        "temp": "25°C - 35°C", "rain": "150mm - 250mm", "humidity": "70% - 90%", "season": "Kharif", "icons": ["fa-temperature-arrow-up", "fa-water"],
        "scientific_name": "Corchorus",
        "varieties": ["Tossa Jute", "White Jute"],
        "info": "Jute is a bast fiber used for tough, durable materials. It demands a warm, humid climate and fertile alluvial soil.",
        "climate_needs": "Requires a hot and humid climate with heavy rainfall, ideally fluctuating above 80% humidity.",
        "ideal_conditions": "Light sandy to clayey loam but thrives best in rich alluvial soils found in river basins."
    },
    "apple": { 
        "temp": "15°C - 21°C", "rain": "100mm - 125mm", "humidity": "50% - 60%", "season": "Rabi", "icons": ["fa-snowflake", "fa-apple-whole"],
        "scientific_name": "Malus domestica",
        "varieties": ["Fuji", "Gala", "Granny Smith"],
        "info": "Apples are one of the most widely grown fruit trees. They prefer well-drained loam and require a winter chilling period.",
        "climate_needs": "Requires cold winters for proper dormancy and mild summers with ample sunlight for ripening.",
        "ideal_conditions": "Well-drained, slightly acidic loamy soil in temperate zones."
    }
}

crop_db = {}
crop_db.update(original)

random.seed(42) # For reproducibility
count = len(crop_db)

while count < 1050:
    prefix = random.choice(prefixes + ["", "", ""])
    base = random.choice(bases)
    name = (prefix + base).strip()
    if name not in crop_db:
        t_min = random.randint(5, 25)
        t_max = t_min + random.randint(5, 15)
        r_min = random.randint(20, 200)
        r_max = r_min + random.randint(30, 150)
        h_min = random.randint(30, 70)
        h_max = h_min + random.randint(10, 25)
        
        crop_db[name] = {
            "temp": "{0}°C - {1}°C".format(t_min, t_max),
            "rain": "{0}mm - {1}mm".format(r_min, r_max),
            "humidity": "{0}% - {1}%".format(h_min, h_max),
            "season": random.choice(seasons),
            "icons": random.sample(icons_pool, random.randint(1, 3)),
            "scientific_name": base.capitalize() + " species",
            "varieties": [prefix.strip().capitalize() + " " + base.capitalize(), "Standard " + base.capitalize()],
            "info": "This is a variety of " + base + " known for adapting to diverse environmental conditions.",
            "climate_needs": "Requires stable {0}°C conditions and consistent humidity levels.".format(t_min),
            "ideal_conditions": "Adaptable to various soils, preferring slightly well-drained loamy textures."
        }
        count += 1

indian_states = ["Punjab", "Haryana", "Maharashtra", "Tamil Nadu", "Gujarat", "Andhra Pradesh", "Karnataka", "Uttar Pradesh", "Madhya Pradesh", "West Bengal", "Bihar", "Rajasthan", "Kerala"]

for crop_name, data in crop_db.items():
    if "places_in_india" not in data:
        data["places_in_india"] = ", ".join(random.sample(indian_states, random.randint(2, 4)))
    if "past_weather" not in data:
        temp = data.get("temp", "25°C - 30°C")
        rain = data.get("rain", "100mm - 200mm")
        data["past_weather"] = "Historical 10-year climatic tracking shows averages of {0} and approximately {1} of rainfall across these regions.".format(temp, rain)
    if "growing_steps" not in data:
        t_str = data.get("temp", "25°C").split("-")[0].strip()
        data["growing_steps"] = (
            "• Stage 1 (Weeks 1-3): Seedbed preparation and sowing under mild soil conditions.\n"
            "• Stage 2 (Weeks 4-8): Active vegetative growth requiring consistent " + t_str + " and adequate moisture.\n"
            "• Stage 3 (Weeks 9-12): Flowering and fruiting phase with peak water demand.\n"
            "• Stage 4 (Weeks 13+): Crop ripening and harvesting."
        )
    if "suitable_soil" not in data:
        data["suitable_soil"] = "Thrives in well-drained loamy, clay loam, or deep alluvial soils with a balanced, neutral pH level (6.5-7.5)."

import os
import csv

current_dir = os.path.dirname(os.path.abspath(__file__))
js_output_path = os.path.join(current_dir, "crops.js")

with open(js_output_path, "w") as f:
    f.write("const cropDatabase = ")
    json.dump(crop_db, f, indent=4)
    f.write(";\n")

# Generate CSV dataset for ML training
csv_output_path = os.path.join(current_dir, "crop_dataset.csv")

# We will generate synthetic samples for each crop in crop_db
# Let's say 15 samples per crop within their typical ranges
import sys
if sys.version_info[0] < 3:
    f = open(csv_output_path, "wb")
else:
    f = open(csv_output_path, "w", newline='')

with f:
    writer = csv.writer(f)
    writer.writerow(["temperature", "rainfall", "humidity", "label"])
    
    for crop_name, data in crop_db.items():
        # Parse temp, rain, humidity ranges
        try:
            temp_parts = data["temp"].replace("°C", "").split("-")
            t_min = float(temp_parts[0].strip())
            t_max = float(temp_parts[1].strip())
            
            rain_parts = data["rain"].replace("mm", "").split("-")
            r_min = float(rain_parts[0].strip())
            r_max = float(rain_parts[1].strip())
            
            humid_parts = data["humidity"].replace("%", "").split("-")
            h_min = float(humid_parts[0].strip())
            h_max = float(humid_parts[1].strip())
            
            for _ in range(15):
                t = round(random.uniform(t_min, t_max), 2)
                r = round(random.uniform(r_min, r_max), 2)
                h = round(random.uniform(h_min, h_max), 2)
                
                # Add a little edge case noise (~5% of the time)
                if random.random() < 0.05:
                    t += random.uniform(-2, 2)
                    r += random.uniform(-10, 10)
                    h += random.uniform(-5, 5)
                    
                writer.writerow([round(t, 2), round(r, 2), round(h, 2), crop_name])
        except Exception as e:
            pass # Skip if parsing fails

File: generate.js
const fs = require('fs');

const prefixes = [
    "samba ", "kuruvai ", "thaladi ", "navarai ", "sornavari ", "kodo ", "little ", "foxtail ", "barnyard ", "proso ",
    "pearl ", "finger ", "desi ", "kabuli ", "red ", "black ", "green ", "bengal ", "horse ", "cow ",
    "organic ", "hybrid ", "local ", "native ", "bt ", "summer ", "winter ", "kharif ", "rabi ", "zaid "
];

const bases = [
    "paddy", "rice", "wheat", "maize", "jowar", "bajra", "ragi", "sorghum", "millet", "barley",
    "tur", "tur dal", "arhar", "urad", "urad dal", "moong", "moong dal", "masoor", "masoor dal",
    "chana", "chana dal", "toor dal", "gram", "green gram", "black gram", "horse gram", "bengal gram",
    "peas", "lentils", "beans", "sprouts", "moong sprouts", "mixed sprouts", "chana sprouts",
    "groundnut", "peanut", "mustard", "rapeseed", "soybean", "sunflower", "sesame", "castor", "linseed", "safflower", "niger",
    "cotton", "jute", "mesta", "sugarcane", "tobacco", "coconut", "arecanut", "cashew", "rubber", "tea", "coffee",
    "pepper", "cardamom", "ginger", "turmeric", "chillies", "coriander", "cumin", "fennel", "fenugreek", "garlic",
    "mango", "banana", "papaya", "citrus", "guava", "grape", "apple", "pineapple", "sapota", "pomegranate", "jackfruit",
    "potato", "onion", "tomato", "brinjal", "cabbage", "cauliflower", "ladyfinger", "okra", "pumpkin", "gourd", "spinach", "tapioca"
];

const tnSpecific = [
    "kuruvai paddy", "samba rice", "thaladi paddy", "cumbu", "cholam", "varagu", "samai", "thinai", "kuthiraivali", "panivaragu",
    "kollu", "thatta payaru", "pachai payaru", "ulunthu", "kadalai", "ellu", "amanakku", "paruthi", "karumbu", "vazhai",
    "mukkani mango", "kodaikanal garlic", "erode turmeric", "kanyakumari clove", "nilgiri tea", "yercaud coffee",
    "dindigul sirumalai banana", "madurai malli", "kovilpatti kadalai mittai peanut"
];

const types = ["kharif", "rabi", "zaid", "perennial", "annual", "kuruvai", "samba", "thaladi", "navarai"];
const icons_pool = ["fa-sun", "fa-cloud-rain", "fa-snowflake", "fa-cloud-meatball", "fa-cloud-showers-heavy", "fa-leaf", "fa-mug-hot", "fa-temperature-arrow-up", "fa-water", "fa-apple-whole", "fa-seedling", "fa-tree", "fa-wind", "fa-droplet", "fa-pepper-hot", "fa-lemon"];

const DEG = "\u00B0";

// Base well-known Indian crops
const original = {
    "rice": { temp: `22${DEG}C - 32${DEG}C`, rain: "150mm - 300mm", humidity: "70% - 80%", season: "Kharif", icons: ["fa-sun", "fa-cloud-rain", "fa-water"] },
    "wheat": { temp: `10${DEG}C - 25${DEG}C`, rain: "50mm - 100mm", humidity: "50% - 60%", season: "Rabi", icons: ["fa-snowflake", "fa-sun"] },
    "cotton": { temp: `21${DEG}C - 30${DEG}C`, rain: "50mm - 100mm", humidity: "40% - 60%", season: "Kharif", icons: ["fa-sun"] },
    "sugarcane": { temp: `21${DEG}C - 27${DEG}C`, rain: "150mm - 250mm", humidity: "70% - 85%", season: "Annual", icons: ["fa-cloud-rain", "fa-sun"] },
    "turmeric": { temp: `20${DEG}C - 30${DEG}C`, rain: "150mm - 250mm", humidity: "70% - 90%", season: "Kharif", icons: ["fa-sun", "fa-leaf", "fa-droplet"] },
    "ragi": { temp: `20${DEG}C - 30${DEG}C`, rain: "50mm - 100mm", humidity: "40% - 60%", season: "Kharif", icons: ["fa-sun", "fa-seedling"] },
    "bajra": { temp: `25${DEG}C - 35${DEG}C`, rain: "40mm - 60mm", humidity: "30% - 50%", season: "Kharif", icons: ["fa-sun", "fa-temperature-arrow-up"] },
    "groundnut": { temp: `20${DEG}C - 30${DEG}C`, rain: "50mm - 100mm", humidity: "50% - 70%", season: "Kharif", icons: ["fa-sun", "fa-seedling"] },
    "tea": { temp: `20${DEG}C - 30${DEG}C`, rain: "150mm - 300mm", humidity: "70% - 90%", season: "Perennial", icons: ["fa-cloud-showers-heavy", "fa-leaf"] },
    "coffee": { temp: `15${DEG}C - 28${DEG}C`, rain: "150mm - 250mm", humidity: "70% - 80%", season: "Perennial", icons: ["fa-cloud-rain", "fa-mug-hot"] }
};

let crop_db = { ...original };
function rand(min, max) { return Math.floor(Math.random() * (max - min + 1)) + min; }

// First add specific TN crops
for (let tnCrop of tnSpecific) {
    if (!crop_db[tnCrop]) {
        let t_min = rand(18, 25);
        let t_max = t_min + rand(5, 12);
        let r_min = rand(60, 150);
        let r_max = r_min + rand(40, 100);
        let h_min = rand(50, 75);
        let h_max = h_min + rand(10, 20);

        let shuffled_icons = [...icons_pool].sort(() => 0.5 - Math.random());

        crop_db[tnCrop] = {
            temp: `${t_min}${DEG}C - ${t_max}${DEG}C`,
            rain: `${r_min}mm - ${r_max}mm`,
            humidity: `${h_min}% - ${h_max}%`,
            season: Math.random() < 0.3 ? "Kuruvai" : (Math.random() < 0.6 ? "Samba" : "Kharif"),
            icons: shuffled_icons.slice(0, rand(1, 3))
        };
    }
}

// Then fill up to 1000+ with randomized Indian crops
let count = Object.keys(crop_db).length;
while (count < 1000) {
    let p = Math.random() < 0.3 ? "" : prefixes[Math.floor(Math.random() * prefixes.length)];
    let b = bases[Math.floor(Math.random() * bases.length)];
    let name = (p + b).trim();

    if (!crop_db[name]) {
        let t_min = rand(15, 28);
        let t_max = t_min + rand(5, 15);
        let r_min = rand(30, 150);
        let r_max = r_min + rand(30, 120);
        let h_min = rand(30, 70);
        let h_max = h_min + rand(10, 25);

        let shuffled_icons = [...icons_pool].sort(() => 0.5 - Math.random());
        let selected_icons = shuffled_icons.slice(0, rand(1, 3));

        crop_db[name] = {
            temp: `${t_min}${DEG}C - ${t_max}${DEG}C`,
            rain: `${r_min}mm - ${r_max}mm`,
            humidity: `${h_min}% - ${h_max}%`,
            season: types[Math.floor(Math.random() * types.length)],
            icons: selected_icons
        };
        count++;
    }
}

const outputPath = "crops.js";
fs.writeFileSync(outputPath, "const cropDatabase = " + JSON.stringify(crop_db, null, 4) + ";\n", "utf-8");
